### PR TITLE
core: optimize dkg protocol

### PR DESCRIPTION
**Pull request by jimmyhu-dexon**
- fast filter already received prv share
- do not propose duplicated complaint

### CODE REVIEW

File: core/dkg-tsig-protocol.go [hunk 1]
```diff
@@ -461,6 +461,18 @@ func (d *dkgProtocol) processPrivateShare(
...
 		return nil
 	}
+	if prvShare.ReceiverID == d.ID {
+		if _, exist := d.prvSharesReceived[prvShare.ProposerID]; exist {
+			return nil
+		}
+	} else {
+		if _, exist := d.antiComplaintReceived[prvShare.ReceiverID]; exist {
+			if _, exist :=
+				d.antiComplaintReceived[prvShare.ReceiverID][prvShare.ProposerID]; exist {
```
Comment (missionliao) [inline core/dkg-tsig-protocol.go]: Would it be better if the result of `d.antiComplaintReceived[prvShare.ReceiverID]` is kept and used in this condition?

Comment (jimmyhu-dexon) [reply core/dkg-tsig-protocol.go]: The code would be very ugly if keeping it.

**Comment by jimmyhu-dexon** [comment]
There would be data race between `processPrivateShare` and `processNackComplaint`. Because a node will propose `antiNackComplaint` but another node might be a little bit slower and not yet construct the `nodesComplaint`

File: core/dkg-tsig-protocol.go (hunk 1 above)
Comment (missionliao) [reply core/dkg-tsig-protocol.go]: I agree that coding style is something very personal, so let it be.

**Comment by missionliao** [comment]
@jimmyhu-dexon As far as I can figure out, both of `processPrivateShare` and `processNackComplaint` are protected by `dkgLock` in `core.configurationChain`. Could you help to elaborate the **date race** you mentioned?

**Comment by jimmyhu-dexon** [comment]
The data race happens between different nodes. A slower node not building `nodesComplained` can start `processPrivateShare` because other nodes are faster.

**Comment by missionliao** [comment]
I still didn't get your point :(
- `nodesComplained` is created in `newDKGProtocol`
- If a slower node start `processPrivateShare` and `nodesComplained` is not created for some reason, we can avoid panic by checking with `if d.nodesComplained != nil && ....`

**Comment by jimmyhu-dexon** [comment]
The timing for creating `nodesComplained` is in `processNackComplaint`

**Comment by missionliao** [comment]
OK, now I now the meaning of `creating nodesComplained` is creating an element in `nodeComplained`. Then if a node start `processPrivateShare` if corresponding element in `nodesComplained` doesn't exist (while other nodes do), what would go wrong?

**Comment by jimmyhu-dexon** [comment]
It will skip processing the private share that is not supposed to be skipped

**Comment by missionliao** [comment]
The filtering by `nodesComplained` I expected is something like:
```
	if prvShare.ReceiverID == d.ID {
		if _, exist := d.prvSharesReceived[prvShare.ProposerID]; exist {
			return nil
		}
	} else {
		if _, exist := d.antiComplaintReceived[prvShare.ReceiverID]; exist {
			if _, exist :=
				d.antiComplaintReceived[prvShare.ReceiverID][prvShare.ProposerID]; exist {
				return nil
			}
		}
		if _, exist := d.nodesComplained[prvShare.ProposerID]; exist {	// the condition I expected
			return nil
		}
	}
```
So if I forbid further private share from a node proposed invalid DKG private, it would break the protocol. But how?

**Comment by jimmyhu-dexon** [comment]
I'm not sure if it's a typo. If a node is got complained, the private share should be processed instead of skipping it.

**Pull request merged**